### PR TITLE
Added Output DataPath rename checks

### PR DESCRIPTION
**Pull request by mmarineBlueQuartz**
* PipelineFilter now checks for renamed outputs between preflights. A signal is emitted when the change in arguments results in a created DataObject having a different DataPath.
* Added IDataCreationAction as a subclass of IDataAction specific for the creation of DataObjects in the DataStructure.
* Pipelines adjust the arguments of PipelineFilter nodes following a renamed DataPath creation.
* Added a renamed output test to PipelineTest.cpp

### CODE REVIEW

**Review by imikejackson — COMMENTED** [review]
I think before this is merged we should have an idea of how NX will use this API to ensure the signals are getting pushed up the chain as needed.

**Review by imikejackson — APPROVED** [review]

File: src/complex/DataStructure/DataPath.cpp [hunk 1]
```diff
@@ -106,6 +106,28 @@ DataPath DataPath::replace(std::string_view symbol, std::string_view targetName)
   return DataPath(std::move(newPath));
 }
 
+bool DataPath::attemptRename(const DataPath& oldPath, const DataPath& newPath)
```
Comment (JDuffeyBQ) [inline src/complex/DataStructure/DataPath.cpp:109]: I'm a little confused on this function. Why does it need to take in `oldPath`?

File: src/complex/Filter/Actions/CreateArrayAction.cpp [hunk 2]
```diff
@@ -77,6 +77,6 @@ const std::vector<usize>& CreateArrayAction::componentDims() const
 
 const DataPath& CreateArrayAction::path() const
 {
-  return m_Path;
+  return getCreatedPath();
```
Comment (JDuffeyBQ) [inline src/complex/Filter/Actions/CreateArrayAction.cpp:80]: Like the warnings say, this is return a reference to a temporary object.

File: src/complex/Filter/Output.hpp [hunk 3]
```diff
@@ -43,6 +43,41 @@ class IDataAction
...
+public:
+  /**
+   * @brief Constructs an IDataCreationAction that takes a DataPath specifying the path to be created.
+   * @param createdPath
+   */
+  IDataCreationAction(const DataPath& createdPath)
+  : m_CreatedPath(createdPath)
+  {
+  }
+  virtual ~IDataCreationAction() = default;
```
Comment (JDuffeyBQ) [inline src/complex/Filter/Output.hpp:61]:
```suggestion
  ~IDataCreationAction() noexcept override = default;
```

File: src/complex/Pipeline/Messaging/OutputRenamedMessage.hpp [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  /**
+   * @brief Constructs a OutputRenamedMessage specifying which AbstractPipelineNode
+   * was added to the target Pipeline and the index it was added at.
+   * @param pipeline
+   * @param newNode
+   * @param index
+   */
+  OutputRenamedMessage(AbstractPipelineNode* pipeline, const PipelineFilter::RenamedPath& renamedOutput);
+
+  virtual ~OutputRenamedMessage();
```
Comment (JDuffeyBQ) [inline src/complex/Pipeline/Messaging/OutputRenamedMessage.hpp:29]:
```suggestion
  ~OutputRenamedMessage() noexcept override;
```

File: src/complex/Pipeline/AbstractPipelineNode.hpp [hunk 5]
```diff
@@ -122,6 +125,16 @@ class COMPLEX_EXPORT AbstractPipelineNode
...
 
+  /**
+   * @brief Attempts to preflight the node using the provided DataStructure.
+   * Returns true if preflighting succeeded. Otherwise, this returns false.
+   * @param data
+   * @param RenamedPaths& renamedPaths = {} Collection of renamed output paths.
+   * Only used for PipelineFilters.
+   * @return bool
+   */
+  virtual bool preflight(DataStructure& data, RenamedPaths& renamedPaths, const std::atomic_bool& shouldCancel) = 0;
```
Comment (JDuffeyBQ) [inline src/complex/Pipeline/AbstractPipelineNode.hpp:136]: If `renamedPaths` is an output, should it be part of the return type?

File: src/complex/Pipeline/PipelineFilter.cpp [hunk 6]
```diff
@@ -168,6 +248,162 @@ bool PipelineFilter::execute(DataStructure& data, const std::atomic_bool& should
...
+ * @brief Removes the overlap between two groups of DataPaths
+ * @param group1
+ * @param group2
+ */
+void removeOverlap(std::vector<DataPath>& group1, std::vector<DataPath>& group2)
+{
+  auto overlap = getOverlap(group1, group2);
+  for(const auto& value : overlap)
+  {
+    std::remove(group1.begin(), group1.end(), value);
```
Comment (JDuffeyBQ) [inline src/complex/Pipeline/PipelineFilter.cpp:308]: If `std::remove` is intending to actually remove items from the vector, you will also need a call to `erase`. This is also the source of the `nodiscard` warnings.

**Review by JDuffeyBQ — CHANGES_REQUESTED** [review]

File: src/complex/DataStructure/DataPath.cpp (hunk 1 above)
Comment (mmarineBlueQuartz) [reply src/complex/DataStructure/DataPath.cpp:109]: This method checks if an attempt to rename is valid before applying it.

Let's say you have DataPath _Foo / Bar / Bazz_. _Foo / Bar_ was renamed to _Foo / Crobar_. _Foo / Bar_ is the old path used to check if the renaming is applicable when renaming. Meanwhile, if _Foo / Random_ was renamed, _Random_ does not match the value in the path and the rename would not be relevant to the current DataPath.

File: src/complex/Pipeline/AbstractPipelineNode.hpp (hunk 5 above)
Comment (mmarineBlueQuartz) [reply src/complex/Pipeline/AbstractPipelineNode.hpp:136]: The output is the success state. The renamedPaths is a parameter that is shared through the course of preflighting a Pipeline.

File: src/complex/Pipeline/Messaging/OutputRenamedMessage.cpp [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+using namespace complex;
+
+OutputRenamedMessage::OutputRenamedMessage(AbstractPipelineNode* pipeline, const PipelineFilter::RenamedPath& renamedOutput)
+: AbstractPipelineMessage(pipeline)
+, m_RenamedPath(renamedOutput)
+{
+}
+
+OutputRenamedMessage::~OutputRenamedMessage() = default;
```
Comment (JDuffeyBQ) [inline src/complex/Pipeline/Messaging/OutputRenamedMessage.cpp:11]:
```suggestion
OutputRenamedMessage::~OutputRenamedMessage() noexcept = default;
```
Fixes exception specification mismatch.

**Review by JDuffeyBQ — CHANGES_REQUESTED** [review]

**Review by JDuffeyBQ — APPROVED** [review]

**Pull request merged**